### PR TITLE
AP-97: Add GraphQL server

### PR DESCRIPTION
**Pull request by chrisdevereux**

### CODE REVIEW

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    this.fetch = fetchFn
+  }
+
+  async request(method: HttpMethod, opts: RequestOpts) {
+    const { headers, body, bodyType = BodyType.Empty, responseType = ResponseType.Empty } = opts
+    const res = await this.fetch(formatUrl(opts.url), {
+      method,
+      body: bodyType.convert(body),
+      headers: toPairs({
+        ...(headers || {}),
```
Comment (richard-hulm) [inline src/lib/plugins/GraphQLPlugin/server/HttpClient.ts]: So in this case, you're not able to override the framework's choice of headers.

If I wanted to have an application/json+customtype

(which is accepted by browsers as application.json) I wouldn't be able to override what the framework had decided should be a application.json header

Can we put 'custom' headers at the bottom, so their value takes precedence

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  get message() {
+    return `Http ${this.status}`
+  }
+
+  get name() {
+    return 'HTTPError'
+  }
+}
+
+export type HttpMethod = 'GET' | 'PUT' | 'PATCH' | 'POST' | 'DELETE'
```
Comment (richard-hulm) [inline src/lib/plugins/GraphQLPlugin/server/HttpClient.ts]: Do we not support others? HEAD / OPTIONS / CONNECT

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  },
+  Empty: {
+    convert: () => undefined,
+    headers: {}
+  }
+}
+
+export const ResponseType = {
+  Json: {
+    convert: (x: Response) => x.json(),
```
Comment (richard-hulm) [inline src/lib/plugins/GraphQLPlugin/server/HttpClient.ts:71]: You may want to handle the case where a 201 is returned

**Review by richard-hulm — DISMISSED** [review]
Couple of minor comments with the HTTP Client,

otherwise, I'm a big fan of the interface for how to write GraphQL resolvers!

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts (hunk 1 above)
Comment (chrisdevereux) [reply src/lib/plugins/GraphQLPlugin/server/HttpClient.ts]: Yeah, explicitly specified headers should probable take precedence 👍

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts (hunk 3 above)
Comment (chrisdevereux) [reply src/lib/plugins/GraphQLPlugin/server/HttpClient.ts:71]: Should we treat 201s as a special case, or put the responsibility on the developer to specify a different response type?

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts (hunk 1 above)
Comment (chrisdevereux) [reply src/lib/plugins/GraphQLPlugin/server/HttpClient.ts]: Although you are able to override the framework's choice of headers by providing a different body/response type.

Comment (richard-hulm) [reply src/lib/plugins/GraphQLPlugin/server/HttpClient.ts]: Yes I saw that afterwards - but even so, for convenience I'd still say users should be able to override it

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts (hunk 3 above)
Comment (richard-hulm) [reply src/lib/plugins/GraphQLPlugin/server/HttpClient.ts:71]: Hmmmm unsure. You could argue either way I think, I'm OK with leaving it as is.

File: src/lib/plugins/GraphQLPlugin/server/HttpClient.ts (hunk 1 above)
Comment (chrisdevereux) [reply src/lib/plugins/GraphQLPlugin/server/HttpClient.ts]: Yeah, that's the behaviour I'd expect

**Review by richard-hulm — APPROVED** [review]

**Pull request merged**